### ISSUE TITLE
Create thumbnail images, pre-sized properly to fit in timeline view

### ISSUE THREAD

**Issue by brentsimmons** [brentsimmons/NetNewsWire#603]
Keep in mind retina issues. Might need to have two sizes of thumbnails for each. And we might need to swap in the proper images when screen resolution changes.

**Comment by brentsimmons** [comment]
Here’s code I used in NetNewsWire Lite 4.0 for making thumbnails:

CGImageRef RSCGImageFromDataWithMaxPixelSize(NSData *imageData, NSInteger maxPixelSize) {
	CGImageSourceRef imageSourceRef = (CGImageSourceRef)[NSMakeCollectable(CGImageSourceCreateWithData((CFDataRef)imageData, NULL)) autorelease];
	if (imageSourceRef == nil)
		return nil;
	size_t numberOfImages = CGImageSourceGetCount(imageSourceRef);
	size_t indexOfImage;
	for (indexOfImage = 0; indexOfImage < numberOfImages; indexOfImage++) {
		NSDictionary *oneImageProperties = [NSMakeCollectable(CGImageSourceCopyPropertiesAtIndex(imageSourceRef, indexOfImage, NULL)) autorelease];
		NSInteger oneImagePixelWidth = [[oneImageProperties objectForKey:(NSString *)kCGImagePropertyPixelWidth] integerValue];
		if (oneImagePixelWidth < 1 || oneImagePixelWidth > maxPixelSize)
			continue;
		NSInteger oneImagePixelHeight = [[oneImageProperties objectForKey:(NSString *)kCGImagePropertyPixelHeight] integerValue];
		if (oneImagePixelHeight > 0 && oneImagePixelHeight <= maxPixelSize) {
			CGImageRef cgImage = CGImageSourceCreateImageAtIndex(imageSourceRef, indexOfImage, NULL);
			if (cgImage != NULL)
				return (CGImageRef)[NSMakeCollectable(cgImage) autorelease];
		}
	}
	return RSCGImageThumbnailFromImageSourceWithMaxPixelSize(imageSourceRef, maxPixelSize);
}


CGImageRef RSCGImageThumbnailFromImageSourceWithMaxPixelSize(CGImageSourceRef imageSourceRef, NSInteger maxPixelSize) {
	NSDictionary *thumbnailOptions = [NSDictionary dictionaryWithObjectsAndKeys:(id)kCFBooleanTrue, (id)kCGImageSourceCreateThumbnailWithTransform, (id)kCFBooleanTrue, (id)kCGImageSourceCreateThumbnailFromImageIfAbsent, [NSNumber numberWithInteger:maxPixelSize], (id)kCGImageSourceThumbnailMaxPixelSize, nil];
	CGImageRef cgImage = CGImageSourceCreateThumbnailAtIndex(imageSourceRef, 0, (CFDictionaryRef)thumbnailOptions);
	if (cgImage == nil)
		return nil;
	return (CGImageRef)[NSMakeCollectable(cgImage) autorelease];
}

**Comment by brentsimmons** [comment]
Okay, all the indentation was lost!

**Comment by brentsimmons** [comment]
Here’s the original: https://github.com/brentsimmons/NetNewsWireLite4/blob/master/RSCore/UI/RSImageUtilities.m

**Comment by vincode-io** [comment]
The way that I implemented this was to only take the main screen scaling resolution into consideration.  Since iOS has 3 different resolutions, I was concerned about computing and caching 3 different versions of the the thumbnails.

I tested this using my non-retina attached monitor and which rendered each avatar at 48x48 pixels and drug it over to my retina MacBook Pro monitor and it looked good.  I didn't notice any degradation in image quality.

I say let's test this implementation and decide if we want a more elaborate one.  I think if we decided we needed to get the resolution change notifications, we could just blow away the caches and recompute at the new resolution rather than always compute for all resolutions.
